File: apps/apple/ClarityCore/Sources/ClarityCore/Recurrence.swift
import Foundation

// RRULE-subset recurrence engine — mirrors packages/shared/src/recurrence.ts.
// Supported: FREQ=DAILY|WEEKLY|MONTHLY|YEARLY, INTERVAL=n,
//            BYDAY=MO,... (weekly), BYMONTHDAY=n (monthly).
// Weekday numbering everywhere: 0 = Monday … 6 = Sunday.

public struct RecurrenceRule: Equatable, Sendable {
    public enum Freq: String, Sendable {
        case daily = "DAILY", weekly = "WEEKLY", monthly = "MONTHLY", yearly = "YEARLY"
    }

    public var freq: Freq
    public var interval: Int
    public var byday: [Int]?
    public var bymonthday: Int?

    public init(freq: Freq, interval: Int = 1, byday: [Int]? = nil, bymonthday: Int? = nil) {
        self.freq = freq
        self.interval = max(1, interval)
        self.byday = byday
        self.bymonthday = bymonthday
    }
}

private let dayCodes = ["MO", "TU", "WE", "TH", "FR", "SA", "SU"]

public func parseRule(_ rule: String) -> RecurrenceRule? {
    var parts: [String: String] = [:]
    let body = rule.hasPrefix("RRULE:") ? String(rule.dropFirst(6)) : rule
    for kv in body.split(separator: ";") {
        let pair = kv.split(separator: "=", maxSplits: 1)
        if pair.count == 2 {
            parts[pair[0].uppercased()] = pair[1].uppercased()
        }
    }
    guard let freqRaw = parts["FREQ"], let freq = RecurrenceRule.Freq(rawValue: freqRaw) else {
        return nil
    }
    var parsed = RecurrenceRule(freq: freq, interval: Int(parts["INTERVAL"] ?? "1") ?? 1)
    if let byday = parts["BYDAY"] {
        let days = byday.split(separator: ",")
            .compactMap { dayCodes.firstIndex(of: $0.trimmingCharacters(in: .whitespaces)) }
        if !days.isEmpty { parsed.byday = Array(Set(days)).sorted() }
    }
    if let dayString = parts["BYMONTHDAY"], let day = Int(dayString), (1...31).contains(day) {
        parsed.bymonthday = day
    }
    return parsed
}

public func formatRule(_ rule: RecurrenceRule) -> String {
    var s = "FREQ=\(rule.freq.rawValue);INTERVAL=\(rule.interval)"
    if let byday = rule.byday, !byday.isEmpty {
        s += ";BYDAY=" + byday.map { dayCodes[$0] }.joined(separator: ",")
    }
    if let bymonthday = rule.bymonthday {
        s += ";BYMONTHDAY=\(bymonthday)"
    }
    return s
}

private var calendar: Calendar {
    var cal = Calendar(identifier: .gregorian)
    cal.firstWeekday = 2 // Monday
    return cal
}

/// 0 = Monday … 6 = Sunday.
private func isoWeekday(_ date: Date) -> Int {
    (calendar.component(.weekday, from: date) + 5) % 7
}

private func startOfDay(_ date: Date) -> Date {
    calendar.startOfDay(for: date)
}

private func startOfWeek(_ date: Date) -> Date {
    calendar.date(byAdding: .day, value: -isoWeekday(date), to: startOfDay(date))!
}

private func addDays(_ date: Date, _ days: Int) -> Date {
    calendar.date(byAdding: .day, value: days, to: date)!
}

/// First occurrence strictly after `after`, anchored at `anchor` (the
/// occurrence being completed). Mirrors nextOccurrence in recurrence.ts.
public func nextOccurrence(
    rule ruleString: String,
    anchor: Date,
    after: Date? = nil
) -> Date? {
    guard let rule = parseRule(ruleString) else { return nil }
    let after = after ?? anchor

    let anchorTime = calendar.dateComponents([.hour, .minute], from: anchor)
    func withAnchorTime(_ day: Date) -> Date {
        calendar.date(
            bySettingHour: anchorTime.hour ?? 0,
            minute: anchorTime.minute ?? 0,
            second: 0,
            of: day
        )!
    }

    if rule.freq == .yearly {
        for i in 1...200 {
            if let candidate = calendar.date(byAdding: .year, value: i * rule.interval, to: anchor),
               candidate > after {
                return candidate
            }
        }
        return nil
    }

    let anchorDay = startOfDay(anchor)
    let anchorWeek = startOfWeek(anchor)
    let base = startOfDay(max(after, anchor))
    var cursor = (withAnchorTime(base) > after && base > anchorDay) ? base : addDays(base, 1)

    for _ in 0..<1600 {
        var matches = false
        switch rule.freq {
        case .daily:
            let diff = calendar.dateComponents([.day], from: anchorDay, to: cursor).day ?? 0
            matches = diff > 0 && diff % rule.interval == 0
        case .weekly:
            let weeks = (calendar.dateComponents([.day], from: anchorWeek, to: startOfWeek(cursor)).day ?? 0) / 7
            let days = rule.byday ?? [isoWeekday(anchorDay)]
            matches = weeks % rule.interval == 0 && days.contains(isoWeekday(cursor))
        case .monthly:
            let months = calendar.dateComponents([.month], from: anchorDay, to: cursor).month ?? 0
            let targetDay = rule.bymonthday ?? calendar.component(.day, from: anchorDay)
            let lastOfMonth = calendar.range(of: .day, in: .month, for: cursor)?.upperBound ?? 32
            let clamped = min(targetDay, lastOfMonth - 1)
            matches = months % rule.interval == 0
                && calendar.component(.day, from: cursor) == clamped
        case .yearly:
            break
        }
        if matches {
            let result = withAnchorTime(cursor)
            if result > after { return result }
        }
        cursor = addDays(cursor, 1)
    }
    return nil
}

public func describeRule(_ ruleString: String) -> String {
    guard let rule = parseRule(ruleString) else { return "custom" }
    let dayNames = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
    func every(_ unit: String) -> String {
        rule.interval == 1 ? "every \(unit)" : "every \(rule.interval) \(unit)s"
    }
    switch rule.freq {
    case .daily: return every("day")
    case .weekly:
        guard let byday = rule.byday, !byday.isEmpty else { return every("week") }
        if byday.count == 7 { return every("day") }
        return every("week") + " on " + byday.map { dayNames[$0] }.joined(separator: ", ")
    case .monthly:
        if let day = rule.bymonthday { return every("month") + " on day \(day)" }
        return every("month")
    case .yearly: return every("year")
    }
}

File: packages/shared/src/recurrence.ts
// A deliberately small RRULE subset (RFC 5545 syntax) that covers the
// recurrences a GTD app needs and is easy to mirror in Swift later.
//
// Supported: FREQ=DAILY|WEEKLY|MONTHLY|YEARLY, INTERVAL=n,
//            BYDAY=MO,TU,... (weekly), BYMONTHDAY=n (monthly).

export interface RecurrenceRule {
  freq: "DAILY" | "WEEKLY" | "MONTHLY" | "YEARLY";
  interval: number;
  byday?: number[]; // 0 = Monday … 6 = Sunday
  bymonthday?: number;
}

const DAY_CODES = ["MO", "TU", "WE", "TH", "FR", "SA", "SU"] as const;
const DAY_NAMES = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"] as const;

export function parseRule(rule: string): RecurrenceRule | null {
  const parts: Record<string, string> = {};
  for (const kv of rule.replace(/^RRULE:/, "").split(";")) {
    const [k, v] = kv.split("=");
    if (k && v) parts[k.toUpperCase()] = v.toUpperCase();
  }
  const freq = parts["FREQ"];
  if (freq !== "DAILY" && freq !== "WEEKLY" && freq !== "MONTHLY" && freq !== "YEARLY") {
    return null;
  }
  const parsed: RecurrenceRule = {
    freq,
    interval: Math.max(1, parseInt(parts["INTERVAL"] ?? "1", 10) || 1),
  };
  if (parts["BYDAY"]) {
    const days = parts["BYDAY"]
      .split(",")
      .map((c) => DAY_CODES.indexOf(c.trim() as (typeof DAY_CODES)[number]))
      .filter((i) => i >= 0);
    if (days.length > 0) parsed.byday = [...new Set(days)].sort();
  }
  if (parts["BYMONTHDAY"]) {
    const d = parseInt(parts["BYMONTHDAY"], 10);
    if (d >= 1 && d <= 31) parsed.bymonthday = d;
  }
  return parsed;
}

export function formatRule(rule: RecurrenceRule): string {
  let s = `FREQ=${rule.freq};INTERVAL=${rule.interval}`;
  if (rule.byday && rule.byday.length > 0) {
    s += `;BYDAY=${rule.byday.map((d) => DAY_CODES[d]).join(",")}`;
  }
  if (rule.bymonthday) s += `;BYMONTHDAY=${rule.bymonthday}`;
  return s;
}

/** 0 = Monday … 6 = Sunday (JS getDay() is 0 = Sunday). */
function isoWeekday(d: Date): number {
  return (d.getDay() + 6) % 7;
}

function startOfDay(d: Date): Date {
  const c = new Date(d);
  c.setHours(0, 0, 0, 0);
  return c;
}

function startOfWeek(d: Date): Date {
  const c = startOfDay(d);
  c.setDate(c.getDate() - isoWeekday(c));
  return c;
}

function addDays(d: Date, n: number): Date {
  const c = new Date(d);
  c.setDate(c.getDate() + n);
  return c;
}

const DAY_MS = 24 * 60 * 60 * 1000;

/**
 * The first occurrence strictly after `after`, anchored at `anchor` (the
 * occurrence being completed — used to phase INTERVAL > 1 and preserve the
 * time of day). Returns null if the rule is unsupported.
 */
export function nextOccurrence(
  rule: string | RecurrenceRule,
  anchor: Date,
  after: Date = anchor
): Date | null {
  const r = typeof rule === "string" ? parseRule(rule) : rule;
  if (!r) return null;

  const withAnchorTime = (day: Date): Date => {
    const c = new Date(day);
    c.setHours(anchor.getHours(), anchor.getMinutes(), 0, 0);
    return c;
  };

  if (r.freq === "YEARLY") {
    for (let i = 1; i <= 200; i++) {
      const candidate = new Date(anchor);
      candidate.setFullYear(anchor.getFullYear() + i * r.interval);
      // setFullYear rolls a Feb 29 anchor into Mar 1 on non-leap years;
      // clamp back to Feb 28 (the Swift engine's Calendar does the same).
      if (candidate.getMonth() !== anchor.getMonth()) candidate.setDate(0);
      if (candidate > after) return candidate;
    }
    return null;
  }

  const anchorDay = startOfDay(anchor);
  const anchorWeek = startOfWeek(anchor);

  // Scan day by day from the later of (anchor, after); include that day
  // itself when the anchor's time of day still lies ahead of `after`.
  const base = startOfDay(after >= anchor ? after : anchor);
  let cursor = withAnchorTime(base) > after && base > anchorDay ? base : addDays(base, 1);

  for (let i = 0; i < 1600; i++, cursor = addDays(cursor, 1)) {
    let matches = false;
    switch (r.freq) {
      case "DAILY": {
        const diff = Math.round((cursor.getTime() - anchorDay.getTime()) / DAY_MS);
        matches = diff > 0 && diff % r.interval === 0;
        break;
      }
      case "WEEKLY": {
        const weeks = Math.round(
          (startOfWeek(cursor).getTime() - anchorWeek.getTime()) / (7 * DAY_MS)
        );
        const days = r.byday ?? [isoWeekday(anchorDay)];
        matches = weeks % r.interval === 0 && days.includes(isoWeekday(cursor));
        break;
      }
      case "MONTHLY": {
        const months =
          (cursor.getFullYear() - anchorDay.getFullYear()) * 12 +
          (cursor.getMonth() - anchorDay.getMonth());
        const targetDay = r.bymonthday ?? anchorDay.getDate();
        const lastOfMonth = new Date(
          cursor.getFullYear(),
          cursor.getMonth() + 1,
          0
        ).getDate();
        matches =
          months % r.interval === 0 &&
          cursor.getDate() === Math.min(targetDay, lastOfMonth);
        break;
      }
    }
    if (matches) {
      const result = withAnchorTime(cursor);
      if (result > after) return result;
    }
  }
  return null;
}

export function describeRule(rule: string | RecurrenceRule): string {
  const r = typeof rule === "string" ? parseRule(rule) : rule;
  if (!r) return "custom";
  const every = (unit: string) =>
    r.interval === 1 ? `every ${unit}` : `every ${r.interval} ${unit}s`;
  switch (r.freq) {
    case "DAILY":
      return every("day");
    case "WEEKLY": {
      const base = every("week");
      if (!r.byday || r.byday.length === 0) return base;
      if (r.byday.length === 7) return every("day");
      return `${base} on ${r.byday.map((d) => DAY_NAMES[d]).join(", ")}`;
    }
    case "MONTHLY":
      return r.bymonthday
        ? `${every("month")} on day ${r.bymonthday}`
        : every("month");
    case "YEARLY":
      return every("year");
  }
}
